File: EPRO.Library.v3.5/EPRO.Library.v3.5/Modules/basReflection.vb


Namespace Modules


    Public Module basReflection


        <System.Runtime.CompilerServices.Extension()> _
        Public Function getAnonymousClassPropValue(pClass As Object, pPropertyName As String,
                                                            Optional ByVal pIndex() As Object = Nothing) As Object

            If pClass Is Nothing Then Return Nothing

            Dim myType As Type = pClass.GetType()
            Dim props As IList(Of System.Reflection.PropertyInfo) = myType.GetProperties()
            Dim fields As IList(Of System.Reflection.FieldInfo) = myType.GetFields()
            If props.Count > 0 Then
                Dim prop = props.Where(Function(x) x.Name = pPropertyName).FirstOrDefault()

                If prop IsNot Nothing Then Return prop.GetValue(pClass, pIndex)
            ElseIf fields.Count > 0 Then
                Dim field = fields.Where(Function(x) x.Name = pPropertyName).FirstOrDefault()

                If field IsNot Nothing Then Return field.GetValue(pClass)

            End If

            Return Nothing

        End Function



        <System.Runtime.CompilerServices.Extension()> _
        Public Function getAnonymousClassMethod(pClass As Object, pMethodName As String) As Reflection.MethodInfo

            If pClass Is Nothing Then Return Nothing

            Dim myType As Type = pClass.GetType()
            Dim meths As IList(Of System.Reflection.MethodInfo) = myType.GetMethods()
            If meths.Count > 0 Then
                Dim meth = meths.Where(Function(x) x.Name = pMethodName).FirstOrDefault()
                If meth IsNot Nothing Then Return meth
            End If

            Return Nothing

        End Function




    End Module


End Namespace

File: EPRO.Library.v3.5/EPRO.Library.v3.5/Modules/basReflection.cs
using System.Collections.Generic;
using System.Linq;

namespace ELibrary.Standard.Modules
{
    public static class basReflection
    {
        public static object getAnonymousClassPropValue(this object pClass, string pPropertyName, object[] pIndex = null)
        {
            if (pClass is null)
                return null;
            var myType = pClass.GetType();
            IList<System.Reflection.PropertyInfo> props = myType.GetProperties();
            IList<System.Reflection.FieldInfo> fields = myType.GetFields();
            if (props.Count > 0)
            {
                var prop = props.Where(x => (x.Name ?? "") == (pPropertyName ?? "")).FirstOrDefault();
                if (prop is object)
                    return prop.GetValue(pClass, pIndex);
            }
            else if (fields.Count > 0)
            {
                var field = fields.Where(x => (x.Name ?? "") == (pPropertyName ?? "")).FirstOrDefault();
                if (field is object)
                    return field.GetValue(pClass);
            }

            return null;
        }

        public static System.Reflection.MethodInfo getAnonymousClassMethod(this object pClass, string pMethodName)
        {
            if (pClass is null)
                return null;
            var myType = pClass.GetType();
            IList<System.Reflection.MethodInfo> meths = myType.GetMethods();
            if (meths.Count > 0)
            {
                var meth = meths.Where(x => (x.Name ?? "") == (pMethodName ?? "")).FirstOrDefault();
                if (meth is object)
                    return meth;
            }

            return null;
        }
    }
}